jib-core/src/main/java/com/google/cloud/tools/jib/event/progress/AllocationCompletionTracker.java

```diff
@@ -0,0 +1,162 @@
+class AllocationCompletionTracker {
+
+  /**
+   * Holds the progress units completed along with a creation order. This is used as the value of
+   * the {@link #completionMap}.
+   */
+  private static class InsertionOrderUnits implements Comparable<InsertionOrderUnits> {
+
+    /** Monotonically-increasing source for {@link #index}. */
+    private static final AtomicInteger currentIndex = new AtomicInteger();
+
+    /** The creation order that monotonically increases. */
+    private final int index = currentIndex.getAndIncrement();
+
+    /**
+     * Progress units completed. This can be shared across multiple threads and should be updated
+     * atomically.
+     */
+    private final AtomicLong units = new AtomicLong();
+
+    /** The {@link Allocation} this is for. */
+    private final Allocation allocation;
+
+    private InsertionOrderUnits(Allocation allocation) {
+      this.allocation = allocation;
+    }
+
+    @Override
+    public int compareTo(InsertionOrderUnits otherInsertionOrderUnits) {
+      return index - otherInsertionOrderUnits.index;
+    }
+  }
+
+  /**
+   * Maps from {@link Allocation} to the number of progress units completed in that {@link
+   * Allocation}, while keeping track of the insertion order of the {@link Allocation}s with {@link
+   * InsertionOrderUnits}.
+   */
+  private final ConcurrentHashMap<Allocation, InsertionOrderUnits> completionMap =
+      new ConcurrentHashMap<>();
+
+  /**
+   * Updates the progress for {@link Allocation} atomically relative to the {@code allocation}.
+   *
+   * <p>For any {@link Allocation}, this method <em>must</em> have been called on all of its parents
+   * beforehand.
+   *
+   * @param allocation the {@link Allocation} to update progress for
+   * @param units the units of progress
+   * @return {@code true} if the map was updated; {@code false} if {@code allocation} was already
+   *     present. Note that there could be false positives if called concurrently, but never false
+   *     negatives.
+   */
+  boolean updateProgress(Allocation allocation, long units) {
+    if (units == 0L) {
+      // Puts the allocation in the map if not present, with progress initialized to 0.
+      boolean alreadyPresent = completionMap.containsKey(allocation);
+      completionMap.computeIfAbsent(allocation, InsertionOrderUnits::new);
+      return !alreadyPresent;
+    }
+
+    completionMap.compute(
+        allocation,
+        (ignored, insertionOrderUnits) -> {
+          if (insertionOrderUnits == null) {
+            insertionOrderUnits = new InsertionOrderUnits(allocation);
+          }
+
+          updateProgress(insertionOrderUnits, units);
+
+          return insertionOrderUnits;
+        });
+    return true;
+  }
+
+  /**
+   * Gets a list of the unfinished {@link Allocation}s in the order in which those {@link
+   * Allocation}s were encountered.
+   *
+   * @return a list of unfinished {@link Allocation}s
+   */
+  List<Allocation> getUnfinishedAllocations() {
```

Comment:
It's not revealed in this PR why we need a list ordered by the time at which events arrived to the event consumer. Can you explain it briefly here? I think it'd be nice and things will become a lot simpler if whoever displays the multiple progress bars in the end would not have to depend on the order of the list, sorted by the time the events arrived.